File: cpp/3903_first_stable_index/first_stable_index.cpp
#include <cassert>
#include <tuple>
#include <vector>

using namespace std;

class Solution {
 public:
  int firstStableIndex(vector<int>& nums, int k) {
    int n = nums.size();
    for (int i = 0; i < n; ++i) {
      int maxValue = nums[i], minValue = nums[i];
      for (int j = 0; j < i; ++j) {
        maxValue = max(maxValue, nums[j]);
      }
      for (int j = i + 1; j < n; ++j) {
        minValue = min(minValue, nums[j]);
      }
      if (maxValue - minValue <= k) {
        return i;
      }
    }
    return -1;
  }
};

int main() {
  vector<tuple<vector<int>, int, int>> tests{
      {{5, 0, 1, 4}, 3, 3},
      {{3, 2, 1}, 1, -1},
      {{0}, 0, 0},
  };

  for (auto& [nums, k, ans] : tests) {
    assert(Solution().firstStableIndex(nums, k) == ans);
  }
}

File: rust/3903_first_stable_index/first_stable_index.rs
struct Solution;

impl Solution {
    pub fn first_stable_index(nums: Vec<i32>, k: i32) -> i32 {
        let n = nums.len();
        for i in 0..n {
            let mut max_value = nums[i];
            let mut min_value = nums[i];
            for j in 0..i {
                max_value = max_value.max(nums[j]);
            }
            for j in i + 1..n {
                min_value = min_value.min(nums[j]);
            }
            if max_value - min_value <= k {
                return i as i32;
            }
        }
        -1
    }
}

fn main() {
    let tests = vec![
        (vec![5, 0, 1, 4], 3, 3),
        (vec![3, 2, 1], 1, -1),
        (vec![0], 0, 0),
    ];

    for (nums, k, ans) in tests {
        assert_eq!(Solution::first_stable_index(nums, k), ans);
    }
}
